core/src/test/java/hudson/util/RetrierTest.java

```diff
@@ -0,0 +1,260 @@
+                // Set the required params
+                (
+                        // action to perform
+                        () -> {
+                            LOG.info("action performed");
+                            return false;
+                        },
+                        // check the result and return true if success
+                        (currentAttempt, result) -> result,
+                        //name of the action
+                        ACTION
+                )
+
+                // Set the optional parameters
+                .withAttempts(ATTEMPTS)
+                .withDelay(100)
+
+                // Construct the object
+                .build();
+
+        // Begin the process
+        Boolean finalResult = r.start();
+
+        Assert.assertFalse(finalResult == null ? false : finalResult);
+
+        String text = Messages.Retrier_NoSuccess(ACTION, ATTEMPTS);
+        assertTrue(String.format("The log should contain '%s'", text), handler.getView().stream().anyMatch(m -> m.getMessage().contains(text)));
+
+    }
+
+    @Test
+    public void failedActionWithExceptionAfterThreeAttemptsWithoutListenerTest() throws Exception {
+        final int ATTEMPTS = 3;
+        final String ACTION = "print";
+
+        RingBufferLogHandler handler = new RingBufferLogHandler(20);
+        Logger.getLogger(Retrier.class.getName()).addHandler(handler);
+
+        Retrier<Boolean> r = new Retrier.Builder<>
+                // Set the required params
+                (
+                        // action to perform
+                        (Callable<Boolean>) () -> {
+                            throw new IndexOutOfBoundsException("Exception allowed considered as failure");
+                        },
+                        // check the result and return true (boolean primitive type) if success
+                        (currentAttempt, result) -> result == null ? false : result,
+                        //name of the action
+                        ACTION
+                )
+
+                // Set the optional parameters
+                .withAttempts(ATTEMPTS)
+                .withDelay(100)
+                .withDuringActionExceptions(new Class[]{IndexOutOfBoundsException.class})
+                // Construct the object
+                .build();
+
+        // Begin the process without catching the allowed exceptions
+        Boolean finalResult = r.start();
+        Assert.assertNull(finalResult);
+
+        String textNoSuccess = Messages.Retrier_NoSuccess(ACTION, ATTEMPTS);
+        assertTrue(String.format("The log should contain '%s'", textNoSuccess), handler.getView().stream().anyMatch(m -> m.getMessage().contains(textNoSuccess)));
+
+        String testException = Messages.Retrier_ExceptionFailed(ATTEMPTS, ACTION);
+        assertTrue(String.format("The log should contain '%s'", testException), handler.getView().stream().anyMatch(m -> m.getMessage().startsWith(testException)));
```

Comment:
Possibly, you could add an assertion checking that between the `build()` and here, there was at least `2x100` milliseconds elapsed? (100 after the first and the second attempt).

Reply:
I've added a new test to check the sleep time works well